archaius2-api/src/main/java/com/netflix/archaius/api/PropertySource.java

```diff
@@ -0,0 +1,142 @@
+package com.netflix.archaius.api;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Contract for a source of raw properties.  A PropertySource may be associated with a single
+ * data source or be a composite of multiple PropertySource instances.  Values may be any object
+ * including strings, boxed primitives, collections, pojos, etc.
+ * 
+ * PropertySource internals are exposed via a set of convenience streaming APIs for common use cases.
+ */
+public interface PropertySource {
+    /**
+     * @return Name of the property source.  This could be an arbitrary name or a file name from whence the
+     * configuration was loaded
+     */
+    String getName();
+
+    /**
+     * @param key
+     * @return Value for the property 
+     */
+    Optional<?> getProperty(String key);
```

Comment:
PropertyNode throws an IllegalArgumentException for an invalid key, but PropertySource returns an Optional<?>.  seems like we need just one strategy